coveo-functools/README.md

```diff
@@ -4,116 +4,312 @@ Introspection, finalizers, delegates, dispatchers, waiters...
+    ...
+
+# breaks static analysis; wrong argument shown for demonstration purposes
+price = calculate_price(SoldTo=dict(Name="Marie"), items={"sku": 123, "price": 19.99})
+```
+
+You could adjust the `Transaction` from earlier class like this:
+
+```python
 @flex
 @dataclass
-class ApiResponse:
-    id: int
-    owner: Owner
-
-# Consider this api response:
-# {
-#     "Id": 1234,
-#     "Owner": {"login": "jonapich"},
-#     "Url": "https://..."
-# }
-response = ApiResponse(**requests.get(...).json)
-assert response.owner.login == 'jonapich'
+class Transaction:
+    sold_to: Customer
+    items: List[SkuItem]
+    rebates: Dict[str, Union[Rebate, List[Rebate]]]
+```
+
+So that you can drop the explicit calls to `flex.deserialize` and use them directly:
+
+```python
+one_transaction = Transaction(**payload)
+list_transactions = [Transaction(**t) for t in [payload, payload]]
+```
+
+
+### `flex` or `deserialize`?
+
+Favor `flex.deserialize` over the decorator pattern:
+- This will make the usages explicit rather than implicit.
+- The additional wrappers created by the decorator may affect performance in the presence of huge structures.
+- You can `flex.deserialize([], hint=List[T])` and get a list, but you cannot `flex(List[T])` directly (both methods demonstrated below)
+
+Generally, it leads to a better design because you end up wiring the `flex.deserialize` call next to the `json.load()` call in a generic manner, and that's 100% of the `flex` code you'll ever need:
+
+
+```python
+class ApiWrapper:
+    def get_transaction(self, id: int) -> Transaction:
+        return self._do_request("GET", f"api/transactions/{id}", hint=Transaction)
+
+    def get_all_transactions(self) -> List[Transaction]:
+        return self._do_request("GET", "api/transactions", hint=List[Transaction])
+  
+    def _do_request(self, method: str, url: str, hint: T) -> T:
+        response = self._session.request(method=method, url=url)
+        return flex.deserialize(response.json, hint=hint)
 ```
 
-In the example above, notice how Owner doesn't have to be decorated?
-This is because @flex works recursively on any type.
-You can decorate it too, but the first call is what matters.
+Because explicit is better than implicit, the above design is generally easier to understand than the one below, where `Transaction` is assumed to be decorated with `@flex`:
 
-#### consideration vs mypy
+```python
+class ApiWrapper:
+    def get_transaction(self, id: int) -> Transaction:
+        return Transaction(**self._do_request("GET", f"api/transactions/{id}"))
+
+    def get_all_transactions(self) -> List[Transaction]:
+        return [Transaction(**data) for data in self._do_request("GET", "api/transactions")]
+  
+    def _do_request(self, method: str, url: str) -> Any:
+        response = self._session.request(method=method, url=url)
+        return response.json
+```
+
+
+## consideration vs mypy

```

Comment:
`vs` means "versus" (in the sense of being compared against), rather than "vis-à-vis". You should reword, especially because it also happens to be a very French phrasing in the first place.

By the way, your lack of upper case in your titles also looks particulary "unenglish" to me, but this is mostly just cosmetics.

Reply:
According to the infinite internet knowledge, vis-à-vis is still a "face to face" kind of concept, I used `Considerations for mypy` instead 🤷🏻‍♂️

Reply:
Yep, it does exist in English. My point was just that you can't switch vis-a-vis with versus, the result doesn't mean the same thing. In any case, `considerations for` is also very French and it probably doesn't mean what you think it does. I was unclear and I apologise -- it was your whole phrasing with _considerations_ that I suggested you reword. Too late now. :P